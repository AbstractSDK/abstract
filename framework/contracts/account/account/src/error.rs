use cosmwasm_std::StdError;

use manager::error::ManagerError;
use proxy::error::ProxyError;
use thiserror::Error;

#[derive(Error, Debug, PartialEq)]
pub enum AccountError {
    #[error("{0}")]
    Std(#[from] StdError),

    // #[error("{0}")]
    // Abstract(#[from] AbstractError),

    // #[error("{0}")]
    // AbstractSdk(#[from] AbstractSdkError),

    // #[error("{0}")]
    // Admin(#[from] AdminError),
    #[error(transparent)]
    Manager(#[from] ManagerError),

    #[error(transparent)]
    Proxy(#[from] ProxyError),
}
