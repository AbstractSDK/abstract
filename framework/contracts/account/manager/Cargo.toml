[package]
name = "abstract-manager"
version = { workspace = true }
authors = { workspace = true }
edition = { workspace = true }
description = "Abstract Manager Contract"
license = { workspace = true }
readme = "README.md"
repository = "https://github.com/AbstractSDK/abstract"

exclude = ["contract.wasm", "hash.txt"]


[lib]
crate-type = ["cdylib", "rlib"]

[features]
default = ["export"]
export = []


[dependencies]
cosmwasm-std = { workspace = true }
cosmwasm-schema = { workspace = true }
cw-controllers = { workspace = true }
cw-storage-plus = { workspace = true }
cw2 = { workspace = true }
cw20 = { workspace = true }
cw-asset = { workspace = true }
schemars = { workspace = true }
serde = { workspace = true }
thiserror = { workspace = true }
abstract-sdk = { workspace = true }
abstract-core = { workspace = true }
semver = { workspace = true }
cw-semver = { workspace = true }
abstract-macros = { workspace = true }
cw-ownable = { workspace = true }

[dev-dependencies]
cw20 = { workspace = true }
abstract-interface = { workspace = true }
cw-orch = { workspace = true }
anyhow = { workspace = true }
account-factory = { workspace = true }
ans-host = { workspace = true }
version-control = { workspace = true }
proxy = { workspace = true }
module-factory = { workspace = true }
rstest = { workspace = true }
speculoos = { workspace = true }
abstract-sdk = { workspace = true, features = ["test-utils"] }
abstract-adapter = { workspace = true, features = ["test-utils"] }
abstract-app = { workspace = true, features = ["test-utils"] }
abstract-testing = { workspace = true }
abstract-macros = { workspace = true }

[profile.release]
rpath = false
lto = true
overflow-checks = true
opt-level = 3
debug = false
debug-assertions = false
codegen-units = 1
panic = 'abort'
incremental = false
