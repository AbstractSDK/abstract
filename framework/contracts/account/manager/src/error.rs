use abstract_core::{
    objects::{validation::ValidationError, version_control::VersionControlError},
    AbstractError,
};
use abstract_sdk::{core::objects::module::ModuleInfo, AbstractSdkError};
use cosmwasm_std::{Instantiate2AddressError, StdError};
use cw_controllers::AdminError;
use thiserror::Error;

#[derive(Error, Debug, PartialEq)]
pub enum ManagerError {
    #[error("{0}")]
    Std(#[from] StdError),

    #[error("{0}")]
    Abstract(#[from] AbstractError),

    #[error("{0}")]
    AbstractSdk(#[from] AbstractSdkError),

    #[error("{0}")]
    Admin(#[from] AdminError),

    #[error("{0}")]
    Validation(#[from] ValidationError),

    #[error("{0}")]
    Ownership(#[from] cw_ownable::OwnershipError),

    #[error("{0}")]
    Instantiate2AddressError(#[from] Instantiate2AddressError),

    #[error("{0}")]
    VersionControlError(#[from] VersionControlError),

    #[error("Module with id: {0} is already installed")]
    ModuleAlreadyInstalled(String),

    #[error("Cannot remove module because {0:?} depend(s) on it.")]
    ModuleHasDependents(Vec<String>),

    #[error("Contract got an unexpected Reply")]
    UnexpectedReply(),

    #[error("The name of the proposed module can not have length 0.")]
    InvalidModuleName {},

    #[error("Registering module fails because caller is not module factory")]
    CallerNotModuleFactory {},

    #[error("A migrate msg is required when when migrating this module")]
    MsgRequired {},

    #[error("{0} not upgradable")]
    NotUpgradeable(ModuleInfo),

    #[error("Cannot migrate {} twice", module_id)]
    DuplicateModuleMigration { module_id: String },

    #[error("Your account is currently suspended")]
    AccountSuspended {},

    #[error("The provided contract version {0} is lower than the current version {1}")]
    OlderVersion(String, String),

    #[error("The provided module {0} was not found")]
    ModuleNotFound(String),

    #[error("Module {module_id} with version {version} does not fit requirement {comp}, post_migration: {post_migration}")]
    VersionRequirementNotMet {
        module_id: String,
        version: String,
        comp: String,
        post_migration: bool,
    },

    #[error("module {0} is a dependency of {1} and is not installed.")]
    DependencyNotMet(String, String),

    #[error("The provided module {0} has an invalid module reference.")]
    InvalidReference(ModuleInfo),

    #[error("Cannot remove proxy")]
    CannotRemoveProxy {},

    #[error("No updates were included")]
    NoUpdates {},

    #[error("invalid configuration action, {}", error)]
    InvalidConfigAction { error: StdError },

    #[error("Must use ProposeOwner to change owner")]
    MustUseProposeOwner {},

    #[error("The address {0} doesn't have an owner, the manager can't determine admin right")]
    NoContractOwner(String),

    #[error("
            Checking the admin recursively failed. 
            You either have the an error in your sub-account configuration or you are not authorized to make this call.
    ")]
    SubAccountAdminVerification,

    #[error("Removing sub account failed")]
    SubAccountRemovalFailed {},

    #[error("Register of sub account failed")]
    SubAccountRegisterFailed {},

    #[error("Can't renounce account, with active sub account")]
    RenounceWithSubAccount {},

    #[error("Can't propose Renounced governance, use update_ownership instead")]
    ProposeRenounced {},

    #[error("Can't create account with Renounced governance")]
    InitRenounced {},
}
