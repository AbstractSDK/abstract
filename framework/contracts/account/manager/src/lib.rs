mod commands;
pub mod contract;
pub mod error;
mod queries;
mod validation;
mod versioning;

#[cfg(test)]
mod test_common {
    use abstract_core::{
        manager,
        objects::{account::AccountTrace, gov_type::GovernanceDetails, AccountId},
    };
    use abstract_testing::prelude::*;
    use cosmwasm_std::{testing::*, DepsMut};

    use crate::contract::ManagerResult;

    /// Initialize the manager with the test owner as the owner
    pub(crate) fn mock_init(mut deps: DepsMut) -> ManagerResult {
        let info = mock_info(TEST_ACCOUNT_FACTORY, &[]);

        crate::contract::instantiate(
            deps.branch(),
            mock_env(),
            info,
            manager::InstantiateMsg {
                account_id: AccountId::new(1, AccountTrace::Local).unwrap(),
                owner: GovernanceDetails::Monarchy {
                    monarch: OWNER.to_string(),
                },
                version_control_address: TEST_VERSION_CONTROL.to_string(),
                module_factory_address: TEST_MODULE_FACTORY.to_string(),
                proxy_addr: TEST_PROXY.to_string(),
                name: "test".to_string(),
                description: None,
                link: None,
                install_modules: vec![],
            },
        )
    }
}
