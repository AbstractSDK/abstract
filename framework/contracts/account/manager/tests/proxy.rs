mod common;
use abstract_adapter::mock::MockExecMsg;
use abstract_core::adapter::AdapterRequestMsg;
use abstract_core::manager::ModuleVersionsResponse;
use abstract_core::module_factory::ModuleInstallConfig;
use abstract_core::objects::fee::FixedFee;
use abstract_core::objects::module::{ModuleInfo, ModuleVersion, Monetization};
use abstract_core::objects::module_reference::ModuleReference;
use abstract_core::objects::namespace::Namespace;
use abstract_core::version_control::UpdateModule;
use abstract_core::{manager::ManagerModuleInfo, PROXY};
use abstract_interface::*;
use abstract_manager::contract::CONTRACT_VERSION;
use abstract_testing::prelude::{TEST_ACCOUNT_ID, TEST_MODULE_ID};
use common::{
    create_default_account, init_mock_adapter, install_adapter, mock_modules, AResult, TEST_COIN,
};
use cosmwasm_std::{coin, to_binary, wasm_execute, Addr, Coin, CosmosMsg};
use cw2::ContractVersion;
use cw_orch::deploy::Deploy;
use cw_orch::prelude::*;
use module_factory::error::ModuleFactoryError;
use speculoos::prelude::*;

#[test]
fn instantiate() -> AResult {
    let sender = Addr::unchecked(common::OWNER);
    let chain = Mock::new(&sender);
    let deployment = Abstract::deploy_on(chain, sender.to_string())?;
    let account = create_default_account(&deployment.account_factory)?;

    let modules = account.manager.module_infos(None, None)?.module_infos;

    // assert proxy module
    assert_that!(&modules).has_length(1);
    assert_that(&modules[0]).is_equal_to(&ManagerModuleInfo {
        address: account.proxy.address()?,
        id: PROXY.to_string(),
        version: cw2::ContractVersion {
            contract: PROXY.into(),
            version: CONTRACT_VERSION.into(),
        },
    });

    // assert manager config
    assert_that!(account.manager.config()?).is_equal_to(abstract_core::manager::ConfigResponse {
        version_control_address: deployment.version_control.address()?,
        module_factory_address: deployment.module_factory.address()?,
        account_id: TEST_ACCOUNT_ID.into(),
        is_suspended: false,
    });
    Ok(())
}

#[test]
fn exec_through_manager() -> AResult {
    let sender = Addr::unchecked(common::OWNER);
    let chain = Mock::new(&sender);
    let deployment = Abstract::deploy_on(chain.clone(), sender.to_string())?;
    let account = create_default_account(&deployment.account_factory)?;

    // mint coins to proxy address
    chain.set_balance(
        &account.proxy.address()?,
        vec![Coin::new(100_000, TEST_COIN)],
    )?;

    // burn coins from proxy
    let proxy_balance = chain
        .app
        .borrow()
        .wrap()
        .query_all_balances(account.proxy.address()?)?;
    assert_that!(proxy_balance).is_equal_to(vec![Coin::new(100_000, TEST_COIN)]);

    let burn_amount: Vec<Coin> = vec![Coin::new(10_000, TEST_COIN)];

    account.manager.exec_on_module(
        cosmwasm_std::to_binary(&abstract_core::proxy::ExecuteMsg::ModuleAction {
            msgs: vec![CosmosMsg::Bank(cosmwasm_std::BankMsg::Burn {
                amount: burn_amount,
            })],
        })?,
        PROXY.to_string(),
        &[],
    )?;

    let proxy_balance = chain
        .app
        .borrow()
        .wrap()
        .query_all_balances(account.proxy.address()?)?;
    assert_that!(proxy_balance).is_equal_to(vec![Coin::new(100_000 - 10_000, TEST_COIN)]);

    Ok(())
}

#[test]
fn default_without_response_data() -> AResult {
    let sender = Addr::unchecked(common::OWNER);
    let chain = Mock::new(&sender);
    let deployment = Abstract::deploy_on(chain.clone(), sender.to_string())?;
    let account = create_default_account(&deployment.account_factory)?;
    let _staking_adapter_one = init_mock_adapter(chain.clone(), &deployment, None)?;

    install_adapter(&account.manager, TEST_MODULE_ID)?;

    chain.set_balance(
        &account.proxy.address()?,
        vec![Coin::new(100_000, TEST_COIN)],
    )?;

    let resp = account.manager.execute_on_module(
        TEST_MODULE_ID,
        Into::<abstract_core::adapter::ExecuteMsg<MockExecMsg>>::into(MockExecMsg),
    )?;
    assert_that!(resp.data).is_none();

    Ok(())
}

#[test]
fn with_response_data() -> AResult {
    let sender = Addr::unchecked(common::OWNER);
    let chain = Mock::new(&sender);
    let deployment = Abstract::deploy_on(chain.clone(), sender.to_string())?;
    let account = create_default_account(&deployment.account_factory)?;
    let staking_adapter = init_mock_adapter(chain.clone(), &deployment, None)?;

    install_adapter(&account.manager, TEST_MODULE_ID)?;

    staking_adapter
        .call_as(&account.manager.address()?)
        .execute(
            &abstract_core::adapter::ExecuteMsg::<MockExecMsg, Empty>::Base(
                abstract_core::adapter::BaseExecuteMsg::UpdateAuthorizedAddresses {
                    to_add: vec![account.proxy.addr_str()?],
                    to_remove: vec![],
                },
            ),
            None,
        )?;

    chain.set_balance(
        &account.proxy.address()?,
        vec![Coin::new(100_000, TEST_COIN)],
    )?;

    let adapter_addr = account
        .manager
        .module_info(TEST_MODULE_ID)?
        .expect("test module installed");
    // proxy should be final executor because of the reply
    let resp = account.manager.exec_on_module(
        cosmwasm_std::to_binary(&abstract_core::proxy::ExecuteMsg::ModuleActionWithData {
            // execute a message on the adapter, which sets some data in its response
            msg: wasm_execute(
                adapter_addr.address,
                &abstract_core::adapter::ExecuteMsg::<MockExecMsg, Empty>::Module(
                    AdapterRequestMsg {
                        proxy_address: Some(account.proxy.addr_str()?),
                        request: MockExecMsg,
                    },
                ),
                vec![],
            )?
            .into(),
        })?,
        PROXY.to_string(),
        &[],
    )?;

    let response_data_attr_present = resp.event_attr_value("wasm-abstract", "response_data")?;
    assert_that!(response_data_attr_present).is_equal_to("true".to_string());

    Ok(())
}

#[test]
fn install_standalone_modules() -> AResult {
    let sender = Addr::unchecked(common::OWNER);
    let chain = Mock::new(&sender);
    let deployment = Abstract::deploy_on(chain.clone(), sender.to_string())?;
    let account = AbstractAccount::new(&deployment, Some(0));

    let standalone1_contract = Box::new(ContractWrapper::new(
        mock_modules::standalone_cw2::mock_execute,
        mock_modules::standalone_cw2::mock_instantiate,
        mock_modules::standalone_cw2::mock_query,
    ));
    let standalone1_id = chain.app.borrow_mut().store_code(standalone1_contract);

    let standalone2_contract = Box::new(ContractWrapper::new(
        mock_modules::standalone_no_cw2::mock_execute,
        mock_modules::standalone_no_cw2::mock_instantiate,
        mock_modules::standalone_no_cw2::mock_query,
    ));
    let standalone2_id = chain.app.borrow_mut().store_code(standalone2_contract);

    // install first standalone
    deployment.version_control.propose_modules(vec![(
        ModuleInfo {
            namespace: Namespace::new("abstract")?,
            name: "standalone1".to_owned(),
            version: ModuleVersion::Version(mock_modules::V1.to_owned()),
        },
        ModuleReference::Standalone(standalone1_id),
    )])?;

    account.install_module(
        "abstract:standalone1",
        &mock_modules::standalone_cw2::MockMsg,
        None,
    )?;

    // install second standalone
    deployment.version_control.propose_modules(vec![(
        ModuleInfo {
            namespace: Namespace::new("abstract")?,
            name: "standalone2".to_owned(),
            version: ModuleVersion::Version(mock_modules::V1.to_owned()),
        },
        ModuleReference::Standalone(standalone2_id),
    )])?;

    account.install_module(
        "abstract:standalone2",
        &mock_modules::standalone_no_cw2::MockMsg,
        None,
    )?;
    Ok(())
}

#[test]
fn install_standalone_versions_not_met() -> AResult {
    let sender = Addr::unchecked(common::OWNER);
    let chain = Mock::new(&sender);
    let deployment = Abstract::deploy_on(chain.clone(), sender.to_string())?;
    let account = AbstractAccount::new(&deployment, Some(0));

    let standalone1_contract = Box::new(ContractWrapper::new(
        mock_modules::standalone_cw2::mock_execute,
        mock_modules::standalone_cw2::mock_instantiate,
        mock_modules::standalone_cw2::mock_query,
    ));
    let standalone1_id = chain.app.borrow_mut().store_code(standalone1_contract);

    // install first standalone
    deployment.version_control.propose_modules(vec![(
        ModuleInfo {
            namespace: Namespace::new("abstract")?,
            name: "standalone1".to_owned(),
            version: ModuleVersion::Version(mock_modules::V2.to_owned()),
        },
        ModuleReference::Standalone(standalone1_id),
    )])?;

    let err = account
        .install_module(
            "abstract:standalone1",
            &mock_modules::standalone_cw2::MockMsg,
            None,
        )
        .unwrap_err();

    if let AbstractInterfaceError::Orch(err) = err {
        let err: ModuleFactoryError = err.downcast()?;
        assert_eq!(
            err,
            ModuleFactoryError::Abstract(abstract_core::AbstractError::UnequalModuleData {
                cw2: mock_modules::V1.to_owned(),
                module: mock_modules::V2.to_owned(),
            })
        );
    } else {
        panic!("wrong error type")
    };

    Ok(())
}

#[test]
fn install_multiple_modules() -> AResult {
    let sender = Addr::unchecked(common::OWNER);
    let chain = Mock::new(&sender);
    chain.add_balance(&sender, vec![coin(86, "token1"), coin(500, "token2")])?;
    let deployment = Abstract::deploy_on(chain.clone(), sender.to_string())?;
    let account = AbstractAccount::new(&deployment, Some(0));

    let standalone1_contract = Box::new(ContractWrapper::new(
        mock_modules::standalone_cw2::mock_execute,
        mock_modules::standalone_cw2::mock_instantiate,
        mock_modules::standalone_cw2::mock_query,
    ));
    let standalone1_id = chain.app.borrow_mut().store_code(standalone1_contract);

    let standalone2_contract = Box::new(ContractWrapper::new(
        mock_modules::standalone_no_cw2::mock_execute,
        mock_modules::standalone_no_cw2::mock_instantiate,
        mock_modules::standalone_no_cw2::mock_query,
    ));
    let standalone2_id = chain.app.borrow_mut().store_code(standalone2_contract);

    // install both standalone
    deployment.version_control.propose_modules(vec![
        (
            ModuleInfo {
                namespace: Namespace::new("abstract")?,
                name: "standalone1".to_owned(),
                version: ModuleVersion::Version(mock_modules::V1.to_owned()),
            },
            ModuleReference::Standalone(standalone1_id),
        ),
        (
            ModuleInfo {
                namespace: Namespace::new("abstract")?,
                name: "standalone2".to_owned(),
                version: ModuleVersion::Version(mock_modules::V1.to_owned()),
            },
            ModuleReference::Standalone(standalone2_id),
        ),
    ])?;

    // add monetization on module1
    let monetization = Monetization::InstallFee(FixedFee::new(&coin(42, "token1")));
    deployment.version_control.update_module_configuration(
        "standalone1".to_owned(),
        Namespace::new("abstract").unwrap(),
        UpdateModule::Versioned {
            version: mock_modules::V1.to_owned(),
            metadata: None,
            monetization: Some(monetization),
            instantiation_funds: None,
        },
    )?;

    // add init funds on module2
    deployment.version_control.update_module_configuration(
        "standalone2".to_owned(),
        Namespace::new("abstract").unwrap(),
        UpdateModule::Versioned {
            version: mock_modules::V1.to_owned(),
            metadata: None,
            monetization: None,
            instantiation_funds: Some(vec![coin(42, "token1"), coin(500, "token2")]),
        },
    )?;

    // Don't allow to attach too much funds
    let err = account
        .install_modules(
            vec![
                ModuleInstallConfig::new(
                    ModuleInfo::from_id_latest("abstract:standalone1")?,
                    Some(to_binary(&mock_modules::standalone_cw2::MockMsg).unwrap()),
                ),
                ModuleInstallConfig::new(
                    ModuleInfo::from_id_latest("abstract:standalone2")?,
                    Some(to_binary(&mock_modules::standalone_no_cw2::MockMsg).unwrap()),
                ),
            ],
            Some(&[coin(86, "token1"), coin(500, "token2")]),
        )
        .unwrap_err();
    assert!(err.root().to_string().contains(&format!(
        "Expected {:?}, sent {:?}",
        vec![coin(84, "token1"), coin(500, "token2")],
        vec![coin(86, "token1"), coin(500, "token2")]
    )));

    // successful install
    account.install_modules_auto(vec![
        ModuleInstallConfig::new(
            ModuleInfo::from_id_latest("abstract:standalone1")?,
            Some(to_binary(&mock_modules::standalone_cw2::MockMsg).unwrap()),
        ),
        ModuleInstallConfig::new(
            ModuleInfo::from_id_latest("abstract:standalone2")?,
            Some(to_binary(&mock_modules::standalone_no_cw2::MockMsg).unwrap()),
        ),
    ])?;

    // Make sure all installed
    let account_module_versions = account.manager.module_versions(vec![
        String::from("abstract:standalone1"),
        // Querying no_cw2 fails
        // String::from("abstract:standalone2"),
    ])?;
    assert_eq!(
        account_module_versions,
        ModuleVersionsResponse {
            versions: vec![
                ContractVersion {
                    contract: String::from(mock_modules::standalone_cw2::MOCK_STANDALONE_ID),
                    version: String::from(mock_modules::V1)
                },
                // ContractVersion {
                //     contract: String::from("abstract:standalone2"),
                //     version: String::from(mock_modules::V1)
                // },
            ]
        }
    );

    let account_module_addresses = account.manager.module_addresses(vec![
        String::from("abstract:standalone1"),
        String::from("abstract:standalone2"),
    ])?;
    let (standalone_addr1, standalone_addr2) = match &account_module_addresses.modules[..] {
        [(_app1, addr1), (_app2, addr2)] => (addr1.clone(), addr2.clone()),
        _ => panic!("bad result from module_addresses"),
    };
    let s1_balance = chain.query_all_balances(&standalone_addr1)?;
    let s2_balance = chain.query_all_balances(&standalone_addr2)?;

    assert!(s1_balance.is_empty());
    assert_eq!(s2_balance, vec![coin(42, "token1"), coin(500, "token2")]);
    Ok(())
}
