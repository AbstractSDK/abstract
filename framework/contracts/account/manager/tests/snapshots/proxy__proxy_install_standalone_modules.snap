---
source: contracts/account/manager/tests/proxy.rs
expression: all_storage
---
"abstract:account-factory":
  - - acseq
    - "1"
  - - cfg
    - "{\"version_control_contract\":\"mock1wug8sewp6cedgkmrmvhl3lf3tulagm9hnvy8p0rppz9yjw0g4wtqwm38hv\",\"ans_host_contract\":\"mock1mzdhwvvh22wrt07w59wxyd58822qavwkx5lcej7aqfkpqqlhaqfsetqc4t\",\"module_factory_address\":\"mock1wkwy0xh89ksdgj9hr347dyd2dw7zesmtrue6kfzyml4vdtz6e5wsldye53\",\"ibc_host\":\"mock14ejqjyq8um4p3xfqj74yld5waqljf88fz25yxnma0cngspxe3les6xzvf5\"}"
  - - contract_info
    - "{\"contract\":\"abstract:account-factory\",\"version\":\"0.20.0\"}"
  - - ownership
    - "{\"owner\":\"mock1pgm8hyk0pvphmlvfjc8wsvk4daluz5tgrw6pu5mfpemk74uxnx9qwrtv4f\",\"pending_owner\":null,\"pending_expiry\":null}"
"abstract:ans-host":
  - - config
    - "{\"next_unique_pool_id\":1}"
  - - contract_info
    - "{\"contract\":\"abstract:ans-host\",\"version\":\"0.20.0\"}"
  - - ownership
    - "{\"owner\":\"mock1pgm8hyk0pvphmlvfjc8wsvk4daluz5tgrw6pu5mfpemk74uxnx9qwrtv4f\",\"pending_owner\":null,\"pending_expiry\":null}"
  - - registered_dexes
    - "[]"
"abstract:ibc-client":
  - - config
    - "{\"version_control\":{\"address\":\"mock1wug8sewp6cedgkmrmvhl3lf3tulagm9hnvy8p0rppz9yjw0g4wtqwm38hv\"},\"ans_host\":{\"address\":\"mock1mzdhwvvh22wrt07w59wxyd58822qavwkx5lcej7aqfkpqqlhaqfsetqc4t\"}}"
  - - contract_info
    - "{\"contract\":\"abstract:ibc-client\",\"version\":\"0.20.0\"}"
  - - module_data
    - "{\"module\":\"abstract:ibc-client\",\"version\":\"0.20.0\",\"dependencies\":[],\"metadata\":null}"
  - - ownership
    - "{\"owner\":\"mock1pgm8hyk0pvphmlvfjc8wsvk4daluz5tgrw6pu5mfpemk74uxnx9qwrtv4f\",\"pending_owner\":null,\"pending_expiry\":null}"
"abstract:ibc-host":
  - - cfg
    - "{\"ans_host\":{\"address\":\"mock1mzdhwvvh22wrt07w59wxyd58822qavwkx5lcej7aqfkpqqlhaqfsetqc4t\"},\"account_factory\":\"mock17p9rzwnnfxcjp32un9ug7yhhzgtkhvl9jfksztgw5uh69wac2pgszycl2y\",\"version_control\":{\"address\":\"mock1wug8sewp6cedgkmrmvhl3lf3tulagm9hnvy8p0rppz9yjw0g4wtqwm38hv\"}}"
  - - contract_info
    - "{\"contract\":\"abstract:ibc-host\",\"version\":\"0.20.0\"}"
  - - ownership
    - "{\"owner\":\"mock1pgm8hyk0pvphmlvfjc8wsvk4daluz5tgrw6pu5mfpemk74uxnx9qwrtv4f\",\"pending_owner\":null,\"pending_expiry\":null}"
"abstract:manager-local-0":
  - - "\u0000\u0007modulesabstract:proxy"
    - "\"mock18dmqux7kjektxhjxlrdxnt6nrvffa3w82l8h0zns9smffvwy47lq0mtca5\""
  - - "\u0000\u0007modulesabstract:standalone1"
    - "\"mock1wr6ur9hmnqgpdh5pc9zd0ad8nxfnm52w5kwdm20k4lnxm3264eusxqqhws\""
  - - "\u0000\u0007modulesabstract:standalone2"
    - "\"mock1v7pmxpk57j5h3sva2dzvepuvg95nzezjcvndwdj5k46ssxwhr24s0l26xn\""
  - - "\u0000{12}is_suspended"
    - "false"
  - - "\u0000{4}info"
    - "{\"name\":\"Default Abstract Account\",\"governance_details\":{\"Monarchy\":{\"monarch\":\"mock1pgm8hyk0pvphmlvfjc8wsvk4daluz5tgrw6pu5mfpemk74uxnx9qwrtv4f\"}},\"chain_id\":\"cosmos-testnet-14002\",\"description\":null,\"link\":null}"
  - - "\u0000{6}config"
    - "{\"version_control_address\":\"mock1wug8sewp6cedgkmrmvhl3lf3tulagm9hnvy8p0rppz9yjw0g4wtqwm38hv\",\"module_factory_address\":\"mock1wkwy0xh89ksdgj9hr347dyd2dw7zesmtrue6kfzyml4vdtz6e5wsldye53\"}"
  - - acc_id
    - "{\"trace\":\"local\",\"seq\":0}"
  - - context
    - "[]"
  - - contract_info
    - "{\"contract\":\"abstract:manager\",\"version\":\"0.20.0\"}"
  - - icontext
    - "[[{\"info\":{\"namespace\":\"abstract\",\"name\":\"standalone2\",\"version\":{\"version\":\"1.0.0\"}},\"reference\":{\"standalone\":10}},\"mock1v7pmxpk57j5h3sva2dzvepuvg95nzezjcvndwdj5k46ssxwhr24s0l26xn\"]]"
  - - ownership
    - "{\"owner\":\"mock1pgm8hyk0pvphmlvfjc8wsvk4daluz5tgrw6pu5mfpemk74uxnx9qwrtv4f\",\"pending_owner\":null,\"pending_expiry\":null}"
"abstract:module-factory":
  - - "\u0000{5}config"
    - "{\"version_control_address\":\"mock1wug8sewp6cedgkmrmvhl3lf3tulagm9hnvy8p0rppz9yjw0g4wtqwm38hv\",\"ans_host_address\":\"mock1mzdhwvvh22wrt07w59wxyd58822qavwkx5lcej7aqfkpqqlhaqfsetqc4t\"}"
  - - contract_info
    - "{\"contract\":\"abstract:module-factory\",\"version\":\"0.20.0\"}"
  - - ownership
    - "{\"owner\":\"mock1pgm8hyk0pvphmlvfjc8wsvk4daluz5tgrw6pu5mfpemk74uxnx9qwrtv4f\",\"pending_owner\":null,\"pending_expiry\":null}"
"abstract:proxy-local-0":
  - - "\u0000{5}state"
    - "{\"modules\":[\"mock1nvqtjjxkdpqtd74zwggnw4nk263uhszps68l8q4g9nz7emttkc3q5lp2w4\",\"mock1wr6ur9hmnqgpdh5pc9zd0ad8nxfnm52w5kwdm20k4lnxm3264eusxqqhws\",\"mock1v7pmxpk57j5h3sva2dzvepuvg95nzezjcvndwdj5k46ssxwhr24s0l26xn\"]}"
  - - "\u0000{6}ans_host"
    - "{\"address\":\"mock1mzdhwvvh22wrt07w59wxyd58822qavwkx5lcej7aqfkpqqlhaqfsetqc4t\"}"
  - - acc_id
    - "{\"trace\":\"local\",\"seq\":0}"
  - - admin
    - "\"mock1nvqtjjxkdpqtd74zwggnw4nk263uhszps68l8q4g9nz7emttkc3q5lp2w4\""
  - - contract_info
    - "{\"contract\":\"abstract:proxy\",\"version\":\"0.20.0\"}"
"abstract:version-control":
  - - "\u0000\u0003lib\u0000\babstract\u0000\u0005proxy0.20.0"
    - "{\"account_base\":6}"
  - - "\u0000\u0003lib\u0000\babstract\u0000\u0007manager0.20.0"
    - "{\"account_base\":5}"
  - - "\u0000\u0003lib\u0000\babstract\u0000\bans-host0.20.0"
    - "{\"native\":\"mock1mzdhwvvh22wrt07w59wxyd58822qavwkx5lcej7aqfkpqqlhaqfsetqc4t\"}"
  - - "\u0000\u0003lib\u0000\babstract\u0000\nibc-client0.20.0"
    - "{\"native\":\"mock1aaf9r6s7nxhysuegqrxv0wpm27ypyv4886medd3mrkrw6t4yfcns7ctvz5\"}"
  - - "\u0000\u0003lib\u0000\babstract\u0000\u000bstandalone11.0.0"
    - "{\"standalone\":9}"
  - - "\u0000\u0003lib\u0000\babstract\u0000\u000bstandalone21.0.0"
    - "{\"standalone\":10}"
  - - "\u0000\u0003lib\u0000\babstract\u0000\u000emodule-factory0.20.0"
    - "{\"native\":\"mock1wkwy0xh89ksdgj9hr347dyd2dw7zesmtrue6kfzyml4vdtz6e5wsldye53\"}"
  - - "\u0000\u0003lib\u0000\babstract\u0000\u000faccount-factory0.20.0"
    - "{\"native\":\"mock17p9rzwnnfxcjp32un9ug7yhhzgtkhvl9jfksztgw5uh69wac2pgszycl2y\"}"
  - - "\u0000\u0003lib\u0000\babstract\u0000\u000fversion-control0.20.0"
    - "{\"native\":\"mock1wug8sewp6cedgkmrmvhl3lf3tulagm9hnvy8p0rppz9yjw0g4wtqwm38hv\"}"
  - - "\u0000\u0004accs\u0000\u0005local\u0000\u0000\u0000\u0000"
    - "{\"manager\":\"mock1nvqtjjxkdpqtd74zwggnw4nk263uhszps68l8q4g9nz7emttkc3q5lp2w4\",\"proxy\":\"mock18dmqux7kjektxhjxlrdxnt6nrvffa3w82l8h0zns9smffvwy47lq0mtca5\"}"
  - - "\u0000\u0004stli\u0000\u0000\u0000\u0000\u0000\u0000\u0000\t"
    - "{\"namespace\":\"abstract\",\"name\":\"standalone1\",\"version\":{\"version\":\"1.0.0\"}}"
  - - "\u0000\u0004stli\u0000\u0000\u0000\u0000\u0000\u0000\u0000\n"
    - "{\"namespace\":\"abstract\",\"name\":\"standalone2\",\"version\":{\"version\":\"1.0.0\"}}"
  - - "\u0000\u0005nmspcabstract"
    - "{\"trace\":\"local\",\"seq\":0}"
  - - "\u0000\u0007nmspc_a\u0000\u0005local\u0000\u0004\u0000\u0000\u0000\u0000abstract"
    - "8"
  - - cfg
    - "{\"account_factory_address\":\"mock17p9rzwnnfxcjp32un9ug7yhhzgtkhvl9jfksztgw5uh69wac2pgszycl2y\",\"allow_direct_module_registration_and_updates\":true,\"namespace_registration_fee\":null}"
  - - contract_info
    - "{\"contract\":\"abstract:version-control\",\"version\":\"0.20.0\"}"
  - - ownership
    - "{\"owner\":\"mock1pgm8hyk0pvphmlvfjc8wsvk4daluz5tgrw6pu5mfpemk74uxnx9qwrtv4f\",\"pending_owner\":null,\"pending_expiry\":null}"

